### PR TITLE
Deprecate the default `EarlyStopping` callback monitor value

### LINKED ISSUE #7894: Remove default monitor value from the EarlyStopping callback
## 🚀 Feature

Remove the default for the Early Stopping callback's monitor: https://github.com/PyTorchLightning/pytorch-lightning/blob/b214442e745e9400e4f842d406939b4e2cd33390/pytorch_lightning/callbacks/early_stopping.py#L91

### Motivation

This is a holdover from pre-1.0 when the Result object defined this key. We should remove the default for these reasons:
- It's possible to construct an EarlyStopping callback without setting this. This can lead to runtime issues during training in case the default key (`early_stop_on`) isn't found in the list of available metrics. Since this happens at the end of either training or validation, this can waste a ton of time compared to failure on construction before training starts.
- We don't want to recommend or suggest that people must log this key within their lightning module to get early stopping to work. `self.log("early_stop_on", <val>)`

### Pitch

Deprecate the default value and force users to explicitly provide a monitor

### PR DESCRIPTION
**Pull request by bamblebam**
## What does this PR do?

Removes the default value for the monitor argument from the early stopping callback. Also adds the deprecation message.
Fixes #7894 

## Before submitting
- [X] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/early_stopping.py [hunk 1]
```diff
@@ -88,7 +88,7 @@ class EarlyStopping(Callback):
 
     def __init__(
         self,
-        monitor: str = 'early_stop_on',
+        monitor: str = None,
```
[the remark this answers is not in the document]
Comment (bamblebam) [reply pytorch_lightning/callbacks/early_stopping.py:91]: Ok will do.

**Comment by bamblebam** [comment]
@ananthsub can you tell me why my test is failing?

File: pytorch_lightning/callbacks/early_stopping.py [hunk 2]
```diff
@@ -88,7 +88,7 @@ class EarlyStopping(Callback):
 
     def __init__(
         self,
-        monitor: str = 'early_stop_on',
+        monitor: str = Optional[None],
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/early_stopping.py:91]: this is invalid python. Optional is a typehint, which needs to be around the str, like this
```suggestion
        monitor: Optional[str] = None,
```

Comment (bamblebam) [reply pytorch_lightning/callbacks/early_stopping.py:91]: Oh alright.

**Review by ananthsub — COMMENTED** [review]
Be sure to add a CHANGELOG entry to the deprecated section of 1.4.0 with this PR:

See here for examples
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md#deprecated

**Comment by bamblebam** [comment]
Hey I updated the changelog.

File: CHANGELOG.md [hunk 3]
```diff
@@ -170,6 +170,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Deprecated `self.log(sync_dist_op)` in favor of `self.log(reduce_fx)`. ([#7891](https://github.com/PyTorchLightning/pytorch-lightning/pull/7891))
 
 
+- Deprecated `monitor` argument in EarlyStopping will be required. ([#7907](https://github.com/PyTorchLightning/pytorch-lightning/pull/7907))
+
```
Comment (ananthsub) [inline CHANGELOG.md:174]:
```suggestion
- Deprecated default value of `monitor` argument in EarlyStopping callback to enforce `monitor` as a required argument ([#7907](https://github.com/PyTorchLightning/pytorch-lightning/pull/7907))

```

**Comment by ananthsub** [comment]
yes this is ready for review

File: pytorch_lightning/callbacks/early_stopping.py [hunk 4]
```diff
@@ -120,6 +119,13 @@ def __init__(
         torch_inf = torch.tensor(np.Inf)
         self.best_score = torch_inf if self.monitor_op == torch.lt else -torch_inf
 
+        if monitor is None:
+            rank_zero_deprecation(
+                "The `monitor` argument will be required to be set starting in v1.6."
```
Comment (carmocca) [inline pytorch_lightning/callbacks/early_stopping.py:124]:
```suggestion
                "The `EarlyStoppinf(monitor)` argument will be required starting in v1.6."
```

**Review by awaelchli — APPROVED** [review]
we may want to consider making the change to our tests wherever the argument is missing.

File: pytorch_lightning/callbacks/early_stopping.py [hunk 5]
```diff
@@ -120,6 +119,13 @@ def __init__(
...
+                " For backward compatibility, setting this to `early_stop_on`."
+            )
+        self.monitor = monitor or 'early_stop_on'
```
[the remark this answers is not in the document]
Comment (bamblebam) [reply pytorch_lightning/callbacks/early_stopping.py:127]: alright I will do the change

[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/callbacks/early_stopping.py:127]: you can just click the button "commit suggestion" here on github below my message. so you don't have to manually commit it. what you prefer!

[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/callbacks/early_stopping.py]: note quoting will get standardized with https://github.com/PyTorchLightning/pytorch-lightning/pull/7783

**Review by carmocca — APPROVED** [review]

**Comment by bamblebam** [comment]
@awaelchli can you tell me why the tests might be failing?

**Review by ananthsub — APPROVED** [review]
great job @bamblebam !

**Comment by bamblebam** [comment]
Thanks for guiding me through this :)
